### PR TITLE
8318124: JFR: Rewrite instrumentation to use Class-File API

### PR DESCRIPTION
**Pull request by egahlin**
Hi,

Could I have a review of an enhancement that replaces the use of ASM with the new Class-File API. This change only deals with bytecode that writes event data into buffers. Bytecode transformations carried out by classes in jdk.jfr.internal.intrument package are kept as is. Plan is to try to replace those with events in java.base.

To simplify the review process, I have tried to keep the code in the classes EventInstrumentation and EventClassBuilder similar to what existed before. Further refactorizations may happen at a later stage. 

Testing: tier1-3 + jdk/jdk/jfr

Thanks
Erik

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318124](https://bugs.openjdk.org/browse/JDK-8318124): JFR: Rewrite instrumentation to use Class-File API (**Enhancement** - P3)

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16195/head:pull/16195` \
`$ git checkout pull/16195`

Update a local copy of the PR: \
`$ git checkout pull/16195` \
`$ git pull https://git.openjdk.org/jdk.git pull/16195/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16195`

View PR using the GUI difftool: \
`$ git pr show -t 16195`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16195.diff">https://git.openjdk.org/jdk/pull/16195.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16195#issuecomment-1763522947)

### CODE REVIEW

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventClassBuilder.java [hunk 1]
```diff
@@ -25,117 +25,119 @@
...
-        buildClassInfo();
-        buildConstructor();
-        buildFields();
-        buildSetMethod();
-        endClass();
-        byte[] bytes = classWriter.toByteArray();
-        ASMToolkit.logASM(fullClassName, bytes);
+        byte[] bytes = Classfile.of().build(ClassDesc.of(fullClassName), cb -> build(cb));
+        try {
+            String name = "/Users/egahlin/DynamicEvent" + (idCounter.longValue() - 1);
```
[the remark this answers is not in the document]
Comment (egahlin) [reply src/jdk.jfr/share/classes/jdk/jfr/internal/EventClassBuilder.java:73]: Yes, looks like I missed some changes. Thanks.

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java [hunk 2]
```diff
@@ -191,18 +198,17 @@ boolean isEnabled() {
...
-                        Object key = values.get(0);
-                        Object value = values.get(1);
-                        if (key instanceof String keyName && value != null) {
-                            if (type == value.getClass()) {
-                                if ("value".equals(keyName)) {
-                                    return (T) value;
+    private static <T> T annotationValue(ClassModel classModel, ClassDesc classDesc, Class<?> type) {
+        String typeDescriptor = classDesc.descriptorString();
+        for (ClassElement ce : classModel.elements()) {
+            if (ce instanceof RuntimeVisibleAnnotationsAttribute rvaa) {
```
[the remark this answers is not in the document]
Comment (egahlin) [reply src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java:203]: I could not get it to work with findAttribute. No annotations were found.

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java [hunk 3]
```diff
@@ -248,11 +252,11 @@ private static List<SettingDesc> buildSettingDescs(Class<?> superClass, ClassMod
                             MethodTypeDesc mtd = m.methodTypeSymbol();
                             if ("Z".equals(mtd.returnType().descriptorString())) {
                                 if (mtd.parameterList().size() == 1) {
-                                    ClassDesc paramType = mtd.parameterList().getFirst();
-                                    if (!paramType.isPrimitive()) {
+                                    ClassDesc type = mtd.parameterList().getFirst();
+                                    if (type.isClassOrInterface()) {
```
Comment (liach) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java:256]: In the original, ASM-based code, there was no such check to ensure the type is non-primitive or non-array. Is this intended?

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (egahlin) [reply src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java:203]: I added logging, and from what I could see, findAttribute always returned Optional.empty().

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java (hunk 3 above)
Comment (egahlin) [reply src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java:256]: Yes.

File: src/jdk.jfr/share/classes/jdk/jfr/internal/util/Bytecode.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+import jdk.internal.classfile.ClassModel;
+import jdk.internal.classfile.Classfile;
+import jdk.internal.classfile.components.ClassPrinter;
+
+/**
+ * Helper class when working with bytecode.
+ */
+public final class Bytecode {
+
+    private final static ClassDesc CD_Thread = classDesc(Thread.class);
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/util/Bytecode.java:45]: Use blessed modifiers order
```suggestion
    private static final ClassDesc CD_Thread = classDesc(Thread.class);
```

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java [hunk 5]
```diff
@@ -213,111 +219,144 @@ private static <T> T annotationValue(ClassNode classNode, String typeDescriptor,
...
-                        String name = m.name;
-                        for (AnnotationNode nameCandidate : m.visibleAnnotations) {
-                            if (ANNOTATION_NAME_DESCRIPTOR.equals(nameCandidate.desc)) {
-                                List<Object> values = nameCandidate.values;
-                                if (values.size() == 1 && values.getFirst() instanceof String s) {
-                                    name = Utils.validJavaIdentifier(s, name);
+        List<SettingDesc> settingDescs = new ArrayList<>();
+        for (MethodModel m : classModel.methods()) {
+            for (MethodElement me : m.elements()) {
+                if (me instanceof RuntimeVisibleAnnotationsAttribute rvaa) {
```
[the remark this answers is not in the document]
Comment (egahlin) [reply src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java:231]: See above.

**Review by mgronlun — APPROVED** [review]

**Pull request closed without merging**